File: 76.最小覆盖子串.go
/*
 * @lc app=leetcode.cn id=76 lang=golang
 *
 * [76] 最小覆盖子串
 */

// @lc code=start
// 滑动窗口
func minWindow(s string, t string) string {
	// Go 里面的字符串其实就是 byte 数组。
	// 每一个 byte 对应一个字符 char。
	ori, cnt := map[byte]int{}, map[byte]int{}
	// 记录 t 中存在的各字符个数。
	for i := 0; i < len(t); i++ {
		ori[t[i]]++
	}

	sLen := len(s)
	minLen := math.MaxInt32
	ansL, ansR := -1, -1

	// 检查当前遍历的字符串是否完全包含了字符串 t。
	check := func() bool {
		for k, v := range ori {
			if cnt[k] < v {
				return false
			}
		}
		return true
	}

	// 这里遍历的是整个 s 字符串
	for l, r := 0, 0; r < sLen; r++ {
		if r < sLen && ori[s[r]] > 0 {
			cnt[s[r]]++
		}
		for check() && l <= r {
			if r-l+1 < minLen {
				minLen = r - l + 1
				ansL, ansR = l, l+minLen
			}
			if _, ok := ori[s[l]]; ok {
				cnt[s[l]] -= 1
			}
			l++
		}
	}
	if ansL == -1 {
		return ""
	}
	return s[ansL:ansR]
}

// @lc code=end

File: 76.最小覆盖子串.java
import java.util.HashMap;
import java.util.Map;

import org.graalvm.compiler.nodes.calc.RightShiftNode;

import jdk.javadoc.internal.doclets.formats.html.resources.standard;

/*
 * @lc app=leetcode.cn id=76 lang=java
 *
 * [76] 最小覆盖子串
 */

// @lc code=start
class Solution {
    private Map<Character, Integer> windowCharCount = new HashMap<>();
    private Map<Character, Integer> tCharCount = new HashMap<>();

    public String minWindow(String s, String t) {
        for (int i = 0; i < t.length(); i++) {
            tCharCount.put(t.charAt(i), tCharCount.getOrDefault(t.charAt(i), 0) + 1);
        }

        int l = 0, r = 0;
        int start = 0, len = Integer.MAX_VALUE;
        while (r < s.length()) {
            char c = s.charAt(r);
            r++;
            windowCharCount.put(c, windowCharCount.getOrDefault(c, 0) + 1);
            while (check()) {
                if (r - l < len) {
                    start = l;
                    len = r - l;
                } 
                char lc = s.charAt(l);
                l++;
                windowCharCount.put(lc, windowCharCount.get(lc) - 1);
            }
        }

        if (len == Integer.MAX_VALUE) {
            return "";
        }
        return s.substring(start, start + len);
    }

    private boolean check() {
        for (Map.Entry<Character, Integer> entry : tCharCount.entrySet()) {
            if (entry.getValue() > windowCharCount.getOrDefault(entry.getKey(), 0)) {
                return false;
            }
        }
        return true;
    }
}
// @lc code=end
